_posts/2022-08-02-stimzi-based-infrastructure-as-a-java-code.md

```diff
@@ -0,0 +1,284 @@
+We can of course wait for the Kafka cluster to be deployed and ready to use from the Java code as well.
+We just need to watch the `.status` section of the custom resource and wait until it says the cluster is ready:
+
+```java
+Crds.kafkaOperation(client).inNamespace(NAMESPACE).withName(NAME).waitUntilCondition(k -> {
+    if (k.getStatus() != null && k.getStatus().getConditions() != null) {
+        return k.getMetadata().getGeneration() == k.getStatus().getObservedGeneration()
+                && k.getStatus().getConditions().stream().anyMatch(c -> "Ready".equals(c.getType()) && "True".equals(c.getStatus()));
+    } else {
+        return false;
+    }
+}, 5, TimeUnit.MINUTES);
+```
+
+The example above will wait for 5 minutes for the cluster to be ready.
+
+### Updating an existing Kafka cluster
+
+The `api` library can be used to modify existing resources as well.
+We need to get the existing custom resource from the Kubernetes API using the client and then we edit it:
+
+```java
+KubernetesClient client = new DefaultKubernetesClient();
+final Kafka updatedKafka = Crds.kafkaOperation(client).inNamespace(NAMESPACE).withName(NAME)
+        .edit(k -> new KafkaBuilder(k)
+                .editSpec()
+                    .editKafka()
+                        .withConfig(Map.of("auto.create.topics.enable", "false"))
+                    .endKafka()
+                .endSpec()
+                .build());
+```
+
+For example in the code snippet above, we change the configuration of the Kafka brokers to disable auto-creation of Kafka topics.
+The code uses the builder interface again to make it easier to edit just one part of the custom resource.
+This code already updates the custom resource inside the Kubernetes API.
+The new YAML corresponding to the Java code would look like this:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: <NAME>
+  namespace: <NAMESPACE>
+spec:
+  kafka:
+    replicas: 3
+    listeners:
+      - name: plain
+        port: 9092
+        type: internal
+        tls: false
+    storage:
+      type: ephemeral
+    config:
+      auto.create.topics.enable: "false"
+  zookeeper:
+    replicas: 3
+    storage:
+      type: ephemeral
+  entityOperator:
+    topicOperator: {}
+    userOperator: {}
+```
+
+Once the resource is updated in Kubernetes, Strimzi will need to roll the Kafka cluster to apply the change.
+We can once again wait for it to complete:
+
+```java
+Crds.kafkaOperation(client).inNamespace(NAMESPACE).withName(NAME).waitUntilCondition(k -> {
+    if (k.getStatus() != null && k.getStatus().getConditions() != null) {
+        return updatedKafka.getMetadata().getGeneration() == k.getStatus().getObservedGeneration()
+                && k.getStatus().getConditions().stream().anyMatch(c -> "Ready".equals(c.getType()) && "True".equals(c.getStatus()));
+    } else {
+        return false;
+    }
```

Comment:
We should encapsulate this and provide it from the `api` module. e.g. `waitUntilCondition(Conditions.reconciled(), 5, TimeUnit.MINUTES)`

Reply:
Maybe we could open a "good start" issue for this, and link to it from the blog post, to see if someone wants to contribute it?

Reply:
Good idea. I opened https://github.com/strimzi/strimzi-kafka-operator/issues/7138 for this.